shotgun_api3/shotgun.py

```diff
@@ -4470,6 +4470,27 @@ def _translate_filters_simple(sg_filter):
     if len(values) == 1 and isinstance(values[0], (list, tuple)):
         values = values[0]
 
+    # Payload optimization: Do not send a full object
+    # just send the `type` and `id` when combining related queries
+    enable_entity_optimization_value = os.environ.get(
+        "SHOTGUN_API_ENABLE_ENTITY_OPTIMIZATION", "0"
+    ).strip().lower()
+    if (
+        enable_entity_optimization_value == "1"
```

Comment:
Another round of suggestions on this block:

```suggestion
    use_entity_optimization = os.environ.get(
        "SHOTGUN_API_ENABLE_ENTITY_OPTIMIZATION", "0"
    ).strip().lower() in ["1", "yes", "true"]

    if (
        use_entity_optimization == "1"
```

This mostly comes from my [previous comment](https://github.com/shotgunsoftware/python-api/pull/360#discussion_r1851141819).

Reply:
Let's enforce a single feature flag value of `1`. The point is to increase performance so I'm dropping a list lookup that is O(n) in favor of a single comparison that I guess is O(1).

Reply:
Thinking some more about it, I would prefer that the setting definition be in the "Config" object so it's only interpreted once:

```diff
diff --git a/shotgun_api3/shotgun.py b/shotgun_api3/shotgun.py
index 99b2a59..bfbe06e 100644
--- a/shotgun_api3/shotgun.py
+++ b/shotgun_api3/shotgun.py
@@ -404,6 +404,10 @@ class _Config(object):
         # In the case that the environment variable is already set, setting the
         # property on the config will override it.
         self.rpc_attempt_interval = 3000
+
+        # TODO
+        self.use_payload_optmiziation = False
+
         # From http://docs.python.org/2.6/library/httplib.html:
         # If the optional timeout parameter is given, blocking operations
         # (like connection attempts) will timeout after that many seconds
@@ -650,6 +654,22 @@ class Shotgun(object):
             raise ValueError("Value of SHOTGUN_API_RETRY_INTERVAL must be positive, "
                              "got '%s'." % self.config.rpc_attempt_interval)
 
+        try:
+            use_payload_optmiziation = os.environ.get(
+                "SHOTGUN_API_ENABLE_ENTITY_OPTIMIZATION"
+            ).strip().lower()
+            if use_payload_optmiziation in ["1", "yes", "true"]:
+                self.config.use_payload_optmiziation = True
+            elif use_payload_optmiziation in ["0", "no", "false"]:
+                self.config.use_payload_optmiziation = False
+            else:
+                LOG.debug(
+                    f'Unrecognized value "{use_payload_optmiziation}" for SHOTGUN_API_ENABLE_ENTITY_OPTIMIZATION.'
+                    ' Sticking to default value ({self.config.use_payload_optmiziation})'
+                )
+        except KeyError:
+            pass
+
         self._connection = None
 
         self.__ca_certs = self._get_certs_file(ca_certs)

```

Reply:
I refactored the code so that the env var is now only read in the `Shotgun` constructor. I didn't use the config object because the filtering is done in functions outside the `Shotgun` class.

I think moving these functions to the `Shotgun` class is risky for now.